tests/lib/linter.js

```diff
@@ -3670,6 +3696,108 @@ describe("Linter", () => {
+                    assert.deepEqual(
+                        types,
+                        ["Program", "ClassDeclaration", "Identifier", "ClassBody", "Decorator", "Identifier"]
+                    );
+                });
+
+                it("eslint-scope should use the visitorKeys (so 'childVisitorKeys.ClassDeclaration' includes 'experimentalDecorators')", () => {
+                    assert(scopeAnalyzeStub.calledOnce);
+                    assert.deepEqual(
+                        scopeAnalyzeStub.firstCall.args[1].childVisitorKeys.ClassDeclaration,
+                        Traverser.DEFAULT_VISITOR_KEYS.ClassDeclaration.concat(["experimentalDecorators"])
+                    );
+                });
+
+                it("should use the same visitorKeys if the source code object is reused", () => {
+                    const types2 = [];
+
+                    linter.defineRule("collect-node-types", () => ({
+                        "*"(node) {
+                            types2.push(node.type);
+                        }
+                    }));
+                    linter.verify(sourceCode, {
+                        rules: {
+                            "collect-node-types": "error"
+                        }
+                    });
+
+                    assert.deepEqual(
+                        types2,
+                        ["Program", "ClassDeclaration", "Identifier", "ClassBody", "Decorator", "Identifier"]
+                    );
+                });
+            });
+
+            describe("if a parser provides 'scope'", () => {
+                let scopeAnalyzeStub = null;
+                let scope = null;
+                let sourceCode = null;
+
+                before(() => {
+                    scopeAnalyzeStub = sinon.spy(escope, "analyze");
+
+                    const parser = path.join(parserFixtures, "enhanced-parser3.js");
+
+                    linter.verify("@foo class A {}", { parser });
+
+                    scope = linter.getScope();
+                    sourceCode = linter.getSourceCode();
+                });
+                after(() => {
+                    scopeAnalyzeStub.restore();
+                });
+
+                it("should not use eslint-scope analyzer", () => {
+                    assert(scopeAnalyzeStub.notCalled);
+                });
+
+                it("should use the scope (so the global scope has the reference of '@foo')", () => {
+                    assert.equal(scope.references.length, 1);
+                    assert.deepEqual(
+                        scope.references[0].identifier.name,
+                        "foo"
+                    );
+                });
+
+                it("should use the same scope if the source code object is reused", () => {
+                    linter.verify(sourceCode, {}, "test.js");
+
+                    assert.equal(linter.getScope(), scope);
```

Comment:
`linter.getScope()` is getting removed in https://github.com/eslint/eslint/pull/9253. I think it would be better to use `context.getScope()`:

```js
linter.defineRule("checker", context => ({
    "Program:exit"() {
        assert.strictEqual(context.getScope(), scope);
    }
}));

linter.verify(sourceCode, { rules: { checker: "error" } });
```

Reply:
Oh, thank you!